ext/nnstreamer/tensor_filter/tensor_filter_llamacpp.cc

```diff
@@ -0,0 +1,350 @@
+ *          application/octet-stream ! filesink location=output.txt
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <nnstreamer_cppplugin_api_filter.hh>
+#include <nnstreamer_log.h>
+#include <nnstreamer_plugin_api_util.h>
+#include <nnstreamer_util.h>
+#include <vector>
+
+#include "ggml.h"
+#include "llama.h"
+/**
+ * @brief Macro for debug mode.
+ */
+#ifndef DBG
+#define DBG FALSE
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+void init_filter_llamacpp (void) __attribute__ ((constructor));
+void fini_filter_llamacpp (void) __attribute__ ((destructor));
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+namespace nnstreamer
+{
+namespace tensorfilter_llamacpp
+{
+/**
+ * @brief Class for TensorFilterLlamaCpp subplugin
+ */
+class TensorFilterLlamaCpp : public tensor_filter_subplugin
+{
+  public:
+  TensorFilterLlamaCpp ();
+  ~TensorFilterLlamaCpp ();
+
+  /* mandatory methods */
+  tensor_filter_subplugin &getEmptyInstance ();
+  void configure_instance (const GstTensorFilterProperties *prop);
+  void invoke (const GstTensorMemory *input, GstTensorMemory *output);
+  void invoke_dynamic (GstTensorFilterProperties *prop,
+      const GstTensorMemory *input, GstTensorMemory *output);
+  void getFrameworkInfo (GstTensorFilterFrameworkInfo &info);
+  int getModelInfo (model_info_ops ops, GstTensorsInfo &in_info, GstTensorsInfo &out_info);
+  int eventHandler (event_ops ops, GstTensorFilterFrameworkEventData &data);
+
+  /* static methods */
+  static void init_filter_llama ();
+  static void fini_filter_llama ();
+
+  private:
+  static TensorFilterLlamaCpp *registered;
+  static const char *name;
+
+  llama_context *ctx;
+  llama_model *model;
+  llama_sampler *smpl;
+  llama_model_params model_params;
+  llama_sampler_chain_params sampler_params;
+  llama_context_params ctx_params;
+};
+
+void init_filter_llama (void) __attribute__ ((constructor));
+void fini_filter_llama (void) __attribute__ ((destructor));
+
+TensorFilterLlamaCpp *TensorFilterLlamaCpp::registered = nullptr;
+const char *TensorFilterLlamaCpp::name = "llamacpp";
+
+/**
+ * @brief Construct a new llamacpp subplugin instance
+ */
+TensorFilterLlamaCpp::TensorFilterLlamaCpp ()
+    : tensor_filter_subplugin (), ctx (nullptr), model (nullptr), smpl (nullptr)
+{
+}
+
+/**
+ * @brief Destructor of TensorFilterllamacpp
+ */
+TensorFilterLlamaCpp::~TensorFilterLlamaCpp ()
+{
+  if (model) {
+    llama_free_model (model);
+    model = nullptr;
+  }
+
+  if (smpl) {
+    llama_sampler_free (smpl);
+    smpl = nullptr;
+  }
+}
+
+/**
+ * @brief Method to get an empty object
+ */
+tensor_filter_subplugin &
+TensorFilterLlamaCpp::getEmptyInstance ()
+{
+  return *(new TensorFilterLlamaCpp ());
+}
+
+/**
+ * @brief Configure llamacpp instance
+ */
+void
+TensorFilterLlamaCpp::configure_instance (const GstTensorFilterProperties *prop)
+{
+  /** @todo: Get ngl value from prop */
+  int ngl = 99;
```

Comment:
How about renaming 'ngl' to 'num_gpu_layers' for code readability?

Reply:
> How about renaming 'ngl' to 'num_gpu_layers' for code readability?

Ditto

> Any plans to bring in the number of layers as a property?

I plan to upload the contents of the prop as a separate PR in the future.
``` /** @todo: Get ngl value from prop */ ```